src/tooltip.tsx

```diff
@@ -112,386 +161,442 @@ const Tooltip: React.FC<Props> = ({
+    } | null>(null);
 
-    const getPosition = (position: Position | undefined) =>
-        isTabletOrSmaller ? position || defaultPositionMobile : position || defaultPositionDesktop;
+    const [arrowComputedProps, setArrowComputedProps] = React.useState<React.CSSProperties>({});
 
-    const [tooltipClientRect, setTooltipClientRect] = React.useState<DOMRect | undefined>(undefined);
+    const targetRef = React.useRef<Element | null>(null);
+    const tooltipRef = React.useRef<HTMLDivElement | null>(null);
+    const [tooltip, setTooltip] = React.useState<HTMLElement | null>(null);
+    const isTouchableDevice = isClientSide() ? window.matchMedia('(pointer: coarse)').matches : false;
 
-    const validatePosition = (position: Position) => {
-        if (!tooltipClientRect) return position;
+    const [isMouseOverTooltip, setIsMouseOverTooltip] = React.useState(false);
+    const [isMouseOverTarget, setIsMouseOverTarget] = React.useState(false);
+    const [isFocused, setIsFocused] = React.useState(false);
+    const [isTooltipOpen, setIsTooltipOpen] = React.useState(false);
+    const isInverse = useIsInverseVariant();
 
-        const validatePositionLeft = (position: Position) => {
-            const hasTopSpace = targetBoundingClientRect.current.top > tooltipClientRect.height;
-            return targetBoundingClientRect.current.left < tooltipClientRect.width
-                ? hasTopSpace
-                    ? 'top'
-                    : 'bottom'
-                : position;
-        };
+    const targetRect = useBoundingRect(targetRef);
+    const windowSize = useWindowSize();
 
-        const alternativePosition = (position: Position) => {
-            return targetBoundingClientRect.current.right + tooltipClientRect.width > window.innerWidth
-                ? validatePositionLeft('left')
-                : position;
-        };
-
-        const positionValidated = {
-            top:
-                targetBoundingClientRect.current.top < tooltipClientRect.height
-                    ? alternativePosition('bottom')
-                    : alternativePosition(position),
-            right: alternativePosition(position),
-            left:
-                targetBoundingClientRect.current.left < tooltipClientRect.width
-                    ? targetBoundingClientRect.current.right + tooltipClientRect.width > window.innerWidth
-                        ? targetBoundingClientRect.current.top < tooltipClientRect.height
-                            ? 'bottom'
-                            : 'top'
-                        : 'right'
-                    : position,
-            bottom:
-                targetBoundingClientRect.current.bottom + tooltipClientRect.height > window.innerHeight
-                    ? alternativePosition('top')
-                    : alternativePosition(position),
-        };
+    React.useEffect(() => {
+        setIsTooltipOpen(tooltipId === openTooltipId);
+    }, [tooltipId, openTooltipId]);
```

Comment:
this is derived state, you don't need this state and this effect, just do this at render time:
```ts
const isTooltipOpen = tooltipId === openTooltipId;
```

Reply:
hmm... check whether this was the cause of the race condition that forced you to set a ref in the provider instead of just the state

Reply:
@pladaria nope, this is not the cause. The problem is the useEffect that calls open/close, it's not related to this one.